src/apps/backend/modules/account/internal/account-writer.ts

```
@@ -64,13 +65,14 @@ export default class AccountWriter {
     newPassword: string,
   ): Promise<Account> {
     const accountHashedPassword = await AccountUtil.hashPassword(newPassword);
-    const dbAccount = await AccountRepository.findByIdAndUpdate(
+    // Type assertion for AccountDB because accountId is already validated to be present in the db
+    const dbAccount = (await AccountRepository.findByIdAndUpdate(
       accountId,
       {
         hashedPassword: accountHashedPassword,
       },
       { new: true },
-    );
+    )) as AccountDB;
```

Is there a way to use generics, override  `findByIdAndUpdate ` by `AccountRepository` such that it automatically returns the right type instead of every consumer uses `as`?

---

@jjalan, ideally, we should add an `if` statement to check whether `dbAccount` is `undefined`. While the parent `function` is already validating it, the write `function` should not rely on external validation. Instead, it should independently ensure that all necessary conditions are met before proceeding.

By adding this check, we can eliminate the need for a TypeScript assertion. TypeScript currently warns that `dbAccount` could be undefined, and we're bypassing this warning with an assertion, assuming the parent `function` has validated it. However, since we have no control over the parent's validation, it's best to handle the check within the writer `function` itself.

---

> it's best to handle the check within the writer function itself.

I see, in that case, it sounds like we should throw if `dbAccount` is not set

---

Yeah, that's correct inside the `if` check we should throw an error.

---

Update via https://github.com/jalantechnologies/boilerplate-mern/pull/198/commits/c70e898aac7eb60557c12ffb3d7edfd703fe9e77